chain/chain/src/chain.rs

```diff
@@ -1903,81 +1903,111 @@ impl Chain {
         Ok(chain_store_update.commit()?)
     }
 
+    pub fn catchup_blocks_step(
+        &mut self,
+        me: &Option<AccountId>,
+        sync_hash: &CryptoHash,
+        blocks_catch_up_state: &mut BlocksCatchUpState,
+        block_catch_up_scheduler: &dyn Fn(BlockCatchUpRequest),
+    ) -> Result<(), Error> {
+        for (queued_block, (saved_store_update, results)) in
+            blocks_catch_up_state.processed_blocks.drain()
+        {
+            match self.block_catch_up_postprocess(&queued_block, results, saved_store_update) {
+                Ok(_) => {
+                    let mut saw_one = false;
+                    for next_block_hash in self.store.get_blocks_to_catchup(&queued_block)?.clone()
+                    {
+                        saw_one = true;
+                        blocks_catch_up_state.pending_blocks.push(next_block_hash.clone());
+                    }
+                    if saw_one {
+                        assert_eq!(
+                            self.runtime_adapter.get_epoch_id_from_prev_block(&queued_block)?,
+                            blocks_catch_up_state.epoch_id
+                        );
+                    }
+                    blocks_catch_up_state.done_blocks.push(queued_block);
+                }
+                Err(_) => {
+                    error!("Error processing block during catch up, retrying");
+                    blocks_catch_up_state.pending_blocks.push(queued_block);
+                }
+            }
+        }
```

Comment:
Is it possible that a malicious attacker can exploit this to cause a node to run out of memory?

Reply:
I am not sure how exactly it can do it. Anyway, I just preserved our current behavior, although I am not sure if it is correct